### PR TITLE
Add more StringSyntaxAttribute syntaxes

### LINKED ISSUE #65634: [API Proposal]: Add more string syntax type constants to StringSyntaxAttribute
EDITED 3/14/2022 by @stephentoub:

```diff
namespace System.Diagnostics.CodeAnalysis
{
    public sealed class StringSyntaxAttribute : Attribute
    {
        public const string DateTimeFormat = nameof(DateTimeFormat);
        public const string Json = nameof(Json);
        public const string Regex = nameof(Regex);

+        public const string CompositeFormat = nameof(CompositeFormat);
+        public const string DateFormat = nameof(DateFormat);
+        public const string EnumFormat = nameof(EnumFormat);
+        public const string GuidFormat = nameof(GuidFormat);
+        public const string NumericFormat = nameof(NumericFormat);
+        public const string TimeFormat = nameof(TimeFormat);
+        public const string TimeSpanFormat = nameof(TimeSpanFormat);
+        public const string Uri = nameof(Uri);
+        public const string Xml = nameof(Xml);
    }
}
```

### Background and motivation

StringSyntaxAttribute defines a few well-known string types. Seems like this could be made considerably more comprehensive in order to standardize on string syntax naming, making it possible to build tooling (like analyzers) around more standard string syntaxes.

### API Proposal

A few ideas:

* CompositeFormat - as in String.Format, etc
* Base64 - happens in config and other places
* Uri
* XML

### API Usage

```C#
[StringSyntax(StringSyntaxAttribute.Base64)]
public string Key { get; set; }

[StringSyntax(StringSyntaxAttribute.Uri)]
public string SourceUri { get; set; }

public string Format([StringSyntax(StringSyntaxAttribute.CompositeFormat)] string format, params object[] args);
```

### Alternative Designs

Without these, there will be a proliferation of names used to represent the same text, which will make it difficult to do tooling that leverages these annotations.

### Risks

_No response_

### PR DESCRIPTION
**Pull request by stephentoub**
Fixes #65634 

cc: @bartonjs

FYI, @CyrusNajmabadi

### CODE REVIEW

File: src/libraries/System.Private.Uri/src/System/Uri.cs [hunk 1]
```diff
@@ -399,7 +399,7 @@ public Uri(Uri baseUri!!, string? relativeUri, bool dontEscape)
         //
         // Uri(string, UriKind);
         //
-        public Uri(string uriString!!, UriKind uriKind)
+        public Uri([StringSyntax(StringSyntaxAttribute.Uri)] string uriString!!, UriKind uriKind)
```
Comment (bartonjs) [inline src/libraries/System.Private.Uri/src/System/Uri.cs:402]: https://github.com/dotnet/runtime/pull/67621/files#diff-947517a24f24fb439159a5c8f47ad9bde9a8c48e5f3ab9672a4fff7a8a73f45cR251 uses `[StringSyntax(StringSyntaxAttribute.Uri, "uriKind")]`.  Assuming the uriKind is relevant to the syntax highlighters/etc, shouldn't the same be done here?

**Review by bartonjs — APPROVED** [review]
One comment about the associated uriKind parameter, otherwise looks reasonable.

The only thing I thought of as non-obvious was Debug.Trace, and you got that.  And you got CryptographicException, which I didn't even think of.  So, seems reasonably complete.

**Comment by stephentoub** [comment]
> And you got CryptographicException, which I didn't even think of

:smile:

File: src/libraries/System.Private.Uri/src/System/Uri.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Private.Uri/src/System/Uri.cs:402]: Yup, good catch.

File: src/libraries/System.Private.Uri/src/System/UriBuilder.cs [hunk 2]
```diff
@@ -25,7 +25,7 @@ public UriBuilder()
         {
         }
 
-        public UriBuilder(string uri)
+        public UriBuilder([StringSyntax(StringSyntaxAttribute.Uri)] string uri)
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Private.Uri/src/System/UriBuilder.cs:28]: That would make sense.  Looks like this goes back a long time.  I'm not going to change it here, but @MihaZupan, fyi.

**Pull request merged**